File: main/base/BinarySearch.py
class Solution:
    @staticmethod
    def binary_search(arr, target):
        low = 0
        high = len(arr) - 1
        while low <= high:
            mid = int(low + (high - low) / 2)
            if arr[mid] == target:
                return mid
            elif arr[mid] > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

    def binary_search_rescurie(self, arr, target, low, high):
        if low < high:
            mid = int(low + (high - low) / 2)
            if arr[mid] == target:
                return mid
            elif arr[mid] < target:
                return self.binary_search_rescurie(arr, target, mid + 1, high)
            else:
                return self.binary_search_rescurie(arr, target, low, mid - 1)
        else:
            return -1

    @staticmethod
    def get_down(arr, target):
        low = 0
        high = len(arr) - 1
        while low < high:
            mid = int(low + (high - low) / 2)
            if arr[mid] < target:
                low = mid + 1
            else:
                high = mid
        return low

    @staticmethod
    def get_up(arr, target):
        low = 0
        high = len(arr) - 1
        while low < high:
            mid = int(low + (high - low) / 2)
            if arr[mid] > target:
                high = mid - 1
            else:
                low = mid
        return low


if __name__ == "__main__":
    arr = [1, 2, 2, 3, 3, 3, 4, 5]
    # print(Solution().binary_search(arr, 3))
    # print(Solution().binary_search(arr, 6))
    # print(Solution().binary_search_rescurie(arr, 5, 0, len(arr)))
    # print(Solution().binary_search_rescurie(arr, 6, 0, len(arr)))
    print(Solution().get_up(arr, 3))
    # Solution().get_up(arr, 3)

File: java/base/BinarySearch.java
package base;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;

public class BinarySearch {

    public static void main(String[] args) {

        int[] arr = {1, 2, 3,3,3,3, 4, 5, 6, 7};
        System.out.println(Arrays.toString(arr));
        System.out.println(binarySearch(arr, 3));

        System.out.println(binarySearch_2(arr, 0, arr.length - 1, 3));
        System.out.println(getDown(arr, 3));
        System.out.println(getUp(arr, 3));


    }

    //二分查找非递归方法
    public static int binarySearch(int[] arr, int num) {

        int low = 0;
        int high = arr.length - 1;
        while (low <= high) {
            int mid = low + ((high - low) / 2);
            if (num == arr[mid])
                return mid;
            else if (num < arr[mid]) {
                high = mid - 1;
            } else
                low = mid + 1;
        }
        return -1;
    }

    //二分查找递归方法
    public static int binarySearch_2(int[] arr, int low, int high, int num) {
        if (low <= high) {
            int mid = (low + high) / 2;
            if (num == arr[mid])
                return mid;
            else if (num < arr[mid])
                return binarySearch_2(arr, low, mid - 1, num);
            else
                return binarySearch_2(arr, mid + 1, high, num);
        } else
            return -1;
    }

    //求数组中下边界问题
    public static int getDown(int[] arr, int k) {
        int l, r;
        for (l = 0, r = arr.length;l<r;){
            int mid = (l+r)/2;
                if(!(arr[mid] >= k)){
                    l = mid+1;
                }else
                    r =mid;
        }
        return l;
    }

    //求数组中上边界问题
    public static int getUp(int[] arr, int k) {
        int l, r;
        for (l = -1, r = arr.length - 1;l<r;){
            int mid = (l+r+1)/2;
            if(!(arr[mid] <= k)){
                r = mid-1;
            }else
                l = mid;
        }
        return l;
    }
}
